solidity/contracts/bridge/WalletCoordinator.sol

```diff
@@ -0,0 +1,542 @@
+// SPDX-License-Identifier: GPL-3.0-only
+
+// ██████████████     ▐████▌     ██████████████
+// ██████████████     ▐████▌     ██████████████
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+// ██████████████     ▐████▌     ██████████████
+// ██████████████     ▐████▌     ██████████████
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+
+pragma solidity 0.8.17;
+
+import {BTCUtils} from "@keep-network/bitcoin-spv-sol/contracts/BTCUtils.sol";
+import {BytesLib} from "@keep-network/bitcoin-spv-sol/contracts/BytesLib.sol";
+import {IWalletRegistry as EcdsaWalletRegistry} from "@keep-network/ecdsa/contracts/api/IWalletRegistry.sol";
+
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+
+import "./BitcoinTx.sol";
+import "./Bridge.sol";
+import "./Deposit.sol";
+import "./Wallets.sol";
+
+/// @title Wallet coordinator.
+/// @notice The wallet coordinator contract aims to facilitate the coordination
+///         of the off-chain wallet members during complex multi-chain wallet
+///         operations like deposit sweeping, redemptions, or moving funds.
+///         Such processes involve various moving parts and many steps that each
+///         individual wallet member must do. Given the distributed nature of
+///         the off-chain wallet software, full off-chain implementation is
+///         challenging and prone to errors, especially byzantine faults.
+///         This contract provides a single and trusted on-chain coordination
+///         point thus taking the riskiest part out of the off-chain software.
+///         The off-chain wallet members can focus on the core tasks and do not
+///         bother about electing a trusted coordinator or aligning internal
+///         states using complex consensus algorithms.
+contract WalletCoordinator is OwnableUpgradeable {
+    using BTCUtils for bytes;
+    using BytesLib for bytes;
+
+    /// @notice Helper structure carrying data necessary to validate the wallet
+    ///         membership of the caller.
+    struct WalletMemberContext {
+        uint32[] walletMembersIDs;
+        uint256 walletMemberIndex;
+    }
+
+    /// @notice Helper structure representing a deposit sweep proposal.
+    ///         Holds the target wallet 20-byte public key hash and candidate
+    ///         deposits that should be part of the sweep.
+    struct DepositSweepProposal {
+        bytes20 walletPubKeyHash;
+        DepositKey[] depositsKeys;
```

Comment:
I tried to avoid bothering the proposal submitter with that. Wallet members can just take the maximum allowed fee from the Bridge and use a % (50%?) of it as a starting point, then bump by another percentage if needed.

Reply:
I think the fee should be published to the chain. The maximum allowed fee from the Bridge is quite large by nature - we do not want to update it too often yet we need to make sure the transaction can be done in case of a fee spike in the network. It is hard to say if 50% or 25% is the right initial value.

Another problem is with bumping the fee. This is not in the scope of the initial implementation but should be done soon after. If the wallet signed a transaction but it's still in the mempool, there should be a mechanism to sign the same data with a higher fee. For the same reasons we coordinate sweep proposition on-chain, we should also coordinate the fee bump on-chain. Now, there are two options:
- Announce the fee bump (more complicated), which requires an on-chain event with a fee.
- Assume it's another sweep proposition (less complicated), also requires the fee to be known on-chain so that we do not start with the same 25% or 50% of the max allowed fee again.

I think the only logic we need is to accept the fee as a parameter, validate it is not higher than the max one, and emit it as a part of the event.

Reply:
See https://github.com/keep-network/tbtc-v2/pull/575/commits/42b167fdc80acb876795432691a3fe9038bbeb0a